### DISCUSSION TITLE
How to start opengrok quickly after index source code

### DISCUSSION THREAD

**Discussion by Andre2020v** [oracle/opengrok#4514] [General]
Dear All,

Firstly I use the command to index source code 
opengrok-indexer -J=-Djava.util.logging.config.file=opengrok/etc/logging.properties -a opengrok/dist/lib/opengrok.jar -- -c /usr/local/bin/ctags -s opengrok/src -d opengrok/data -H -P -S -G -W opengrok/etc/configuration.xml -U http://localhost:8080/source

then I reboot the linux server ,  how to restart the opengrok ?  

Best wishes,
Andrew

**Comment by vladak** [comment]
What exactly do you mean by restarting ? When a system is rebooted, all the OpenGrok processes (webapp and also indexer, if running at the moment) will be terminated.

**Reply by Andre2020v** [reply]
The hardware ( the linux server) is rebooted.  e.g. after linux system updates some patch.  I hope there are some command to start the opengrok quickly ,not index source code again.  because the code is not changed anything , index source code spends a few hours .

**Reply by vladak** [reply]
What OpenGrok version are you running ? What is the structure (types of repositories used, if any) and size of the source root directory ?

**Reply by Andre2020v** [reply]
Thanks for your reply.

1.  I use 1.12.27 , the latest version.

2. I don't use any version control systems like SVN, git,  becasue it is used for personal studying android code.   In fact , firstly I  keep git repository ， but I find it will be terrible size . e.g.  it is 120GB  that source code includes git repository , doing the index , the data dir of opengrok will reach 1TB and spend 2 days to index code . so firstly I remove the git repository , only index the source without git repository , it is faster and spend less disk space

3.  It is AOSP( Android Open Source Project) ,  code size is about  59GB without git repository 

because the linux server is my personal pc installing ubuntu 22.04, so I hope afte study code , I can turn off the pc . next time , I can quickly start opengrok and load the index data to search for full text, definitions, symbols ... But I'm not sure if my idea is feasible

**Reply by vladak** [reply]
In such scenario, i.e. without source code management system [*], the indexer has to traverse the whole directory structure and check last modified time for each found file against its respective document stored in the index, if any. This is still much faster than reindex from scratch. With system I/O caches not warmed up (as it happens to be after reboot) and possibly slower I/O subsystem, the time spent in this operation might be significant.

But I think you are talking about webapp startup, not running the indexer. The webapp startup should be usually quick.

[*]  with the input data being stored in a repository, allows the indexer to quick get the list of changes since the last time the indexer has run. This is called _history based reindex_. This is currently supported for Mercurial and Git - with certain caveats like all repositories for given project have to support this for this indexing method to be used.

**Reply by Andre2020v** [reply]
Thanks for your help !

Sorry for making your confuse , because I'm not familiar with java , webapp ....

now let's me show process

1. index code

- cd tomcat10/bin ; run ./start.sh

- . ./env/bin/activate (for using python)

- opengrok-indexer -J=-Djava.util.logging.config.file=opengrok/etc/logging.properties -a opengrok/dist/lib/opengrok.jar -- -c /usr/local/bin/ctags -s opengrok/src -d opengrok/data -H -P -S -G -W opengrok/etc/configuration.xml -U http://localhost:8080/source

- open a Web Browser ,then input "http://localhost:8080/source" to use opengrok to study code.

- turn off PC

2. webapp startup

- turn on PC

- cd tomcat/bin ; run ./start.sh

- . ./env/bin/activate ( for using python )

- opengrok-deploy -c opengrok/etc/configuration.xml  opengrok/dist/lib/source.war tomcat10/webapps

"But I think you are talking about webapp startup, not running the indexer" , it made me suddenly realize. The answer I want is "opengrok-deploy -c opengrok/etc/configuration.xml opengrok/dist/lib/source.war tomcat10/webapps" .

one more question, if opengrok-indexer can controls the index size that there is git repository , or uses git to only download a few branch to reduce the size of git repository ?

**Reply by vladak** [reply]
Once the initial index is created (the `opengrok-indexer` command in step 1 in your case), the webapp can be started anytime without running the indexer. The webapp might still do some heavier processing to generated suggester data, however that happens in the background.

OpenGrok itself is mostly agnostic w.r.t. input data. For instance, it will not perform perform Git clone or such. Now, there is some tooling distributed with OpenGrok that will help with repository synchronization, however the indexer itself merely processes the data available in source root. Same for webapp with some very minor exceptions (e.g. when getting historical version of a file).

**Comment by Andre2020v** [comment]
Thanks for your help !  I'm very happy that you can reply to my questions.
